### PR TITLE
Сoncrete values for enums

### PR DESCRIPTION
**Pull request by Damtev**
Closes https://github.com/UnitTestBot/UTBotJava/issues/18, https://github.com/UnitTestBot/UTBotJava/issues/32

To increase the performance of symbolic analysis for source code that uses enums, we can extract these enums concretely from runtime instead of enum's initializers analysis which can require a lot of time for enums with a lot of constants (because enum's static initializer fills fields for every enum constant, creates `$VALUES` array, invokes `java.lang.Enum` initializer for every constant, etc). 

But very important thing is that we cannot process concretely every enum but only  **isolated** - enums that do not affect the system state (like other classes' statics) in their initializers and do not use external statics to initialize its own statics. To prevent concrete processing of enum that **is not isolated**, we need to traverse its static initializer and check that it does not initialize external statics and does not use external statics to initialize its own statics.

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+    alreadyProcessed += this
+
+    // active body could be missing (<java.lang.Enum: void <init>(java.lang.String,int)>, for example)
+    // so consider it as not affecting external statics
+    if (!canRetrieveBody()) {
+        return false
+    }
+
+    return activeBody.units.any {
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt:194]: Above you checked that you could retrieve a body, but didn't do it. You might do not have this body here yet

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+        when (it) {
+            is JAssignStmt -> {
+                val leftOp = it.leftOp
+
+                if (leftOp !is StaticFieldRef) {
+                    return@any false
+                }
+
+                val declaringClass = leftOp.fieldRef.declaringClass()
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt:207]: Same about declaringClass

**Review by CaelmBleidd — CHANGES_REQUESTED** [review]
Add a design doc for concrete processing of enum's clinit sections

File: utbot-framework/src/main/kotlin/org/utbot/engine/Resolver.kt [hunk 3]
```diff
@@ -1141,6 +1143,29 @@ private fun UtBotSymbolicEngine.arrayToMethodResult(
     )
 }
 
+fun UtBotSymbolicEngine.constructEnumStaticFieldResult(
+    fieldName: String,
+    fieldType: Type,
+    declaringClass: SootClass,
+    enumConstantSymbolicResultsByName: Map<String, MethodResult>,
+    existingEnumSymbolicValue: Any?,
```
[the remark this answers is not in the document]
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Resolver.kt]: This method is quite properly documented in its body, so I just fixed the wrong parameter's name

File: utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt [hunk 4]
```diff
@@ -922,7 +928,9 @@ class UtBotSymbolicEngine(
             // Note that this list is not exhaustive, so it may be supplemented in the future.
             val packagesToProcessConcretely = javaPackagesToProcessConcretely + sunPackagesToProcessConcretely
 
-            return packagesToProcessConcretely.any { className.startsWith(it) }
+            val declaringClass = fieldRef.fieldRef.declaringClass()
```
[the remark this answers is not in the document]
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt]: Fixed

File: utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt [hunk 5]
```diff
@@ -963,26 +978,123 @@ class UtBotSymbolicEngine(
...
+        //      ClassWithEnum$StatusEnum[] $r0, $r2;
+        //      java.lang.Object $r1;
+        //      $r0 = <ClassWithEnum$StatusEnum: ClassWithEnum$StatusEnum[] $VALUES>;
+        //      $r1 = virtualinvoke $r0.<java.lang.Object: java.lang.Object clone()>();
+
+        // so, we have to make an update for the local $r0
+        if (stmt is JAssignStmt) {
+            val local = stmt.leftOp as JimpleLocal
+            val localUpdate = localMemoryUpdate(
+                LocalVariable(local.name) to curFieldSymbolicValueForLocalVariable
```
[the remark this answers is not in the document]
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt]: Fixed

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt (hunk 1 above)
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt]: Changed to jimpleBody

File: utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt (hunk 2 above)
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/engine/util/statics/concrete/EnumConcreteUtils.kt]: Fixed

**Review by CaelmBleidd — APPROVED** [review]

**Pull request merged**